src/libtrx/game/items.c

```diff
@@ -690,3 +768,103 @@ int32_t Item_Explode(
 
     return !(item->mesh_bits & (0x7FFFFFFF >> (31 - obj->mesh_count)));
 }
+
+ANIM_FRAME *Item_GetBestFrame(const ITEM *const item)
+{
+    ANIM_FRAME *frames[2];
+    int32_t rate = 0;
+    const int32_t frac = Item_GetFrames(item, frames, &rate);
+    return frames[(frac > rate / 2) ? 1 : 0];
+}
+
+const BOUNDS_16 *Item_GetBoundsAccurate(const ITEM *const item)
+{
+    int32_t rate;
+    ANIM_FRAME *frames[2];
+    const int32_t frac = Item_GetFrames(item, frames, &rate);
+    if (frames[0] == nullptr) {
+        return &m_NullBounds;
+    }
+
+    if (frac == 0) {
+        return &frames[0]->bounds;
+    }
+
+#define CALC(target, b1, b2, prop)                                             \
+    target->prop = (b1)->prop + ((((b2)->prop - (b1)->prop) * frac) / rate);
+
+    BOUNDS_16 *const result = &m_InterpolatedBounds;
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, min.x);
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, max.x);
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, min.y);
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, max.y);
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, min.z);
+    CALC(result, &frames[0]->bounds, &frames[1]->bounds, max.z);
+#undef CALC
+    return result;
+}
+
+int32_t Item_GetFrames(const ITEM *item, ANIM_FRAME *frames[], int32_t *rate)
+{
+    const ANIM *const anim = Item_GetAnim(item);
+    if (anim->frame_ptr == nullptr) {
+        frames[0] = nullptr;
+        return 0;
+    }
+
+    const int32_t cur_frame_num = item->frame_num - anim->frame_base;
+    const int32_t last_frame_num = anim->frame_end - anim->frame_base;
+    const int32_t key_frame_span = anim->interpolation;
+    const int32_t key_frame_shift = cur_frame_num % key_frame_span;
+    const int32_t first_key_frame_num = cur_frame_num / key_frame_span;
+    const int32_t second_key_frame_num = first_key_frame_num + 1;
+
+    const int32_t numerator = key_frame_shift;
+    int32_t denominator = key_frame_span;
+    if (numerator != 0) {
+#if TR_VERSION == 1
+        if (second_key_frame_num > anim->frame_end) {
+            denominator =
+                anim->frame_end + key_frame_span - second_key_frame_num;
+        }
+#else
+        const int32_t second_key_frame_num2 =
+            (cur_frame_num / key_frame_span + 1) * key_frame_span;
+        if (second_key_frame_num2 > anim->frame_end) {
+            denominator += anim->frame_end - second_key_frame_num2;
+        }
+#endif
```

Comment:
It'd be nice to consolidate this. I wonder if it's possible?

Reply:
I initially tried TR1 using TR2's code here and, while I only played around Vilcabamba for a little bit, it seemed fine. We could try it separately I think and see how we get on.